File: app/src/main/java/org/cru/godtools/activity/ToolDetailsActivity.java
package org.cru.godtools.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import org.cru.godtools.R;
import org.cru.godtools.analytics.model.AnalyticsScreenEvent;
import org.cru.godtools.fragment.ToolDetailsFragment;
import org.cru.godtools.model.Tool;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import static org.cru.godtools.analytics.model.AnalyticsScreenEvent.SCREEN_TOOL_DETAILS;
import static org.cru.godtools.base.Constants.EXTRA_TOOL;

public class ToolDetailsActivity extends BasePlatformActivity implements ToolDetailsFragment.Callbacks {
    private static final String TAG_MAIN_FRAGMENT = "mainFragment";

    @Nullable
    private /*final*/ String mTool = Tool.INVALID_CODE;

    public static void start(@NonNull final Activity context, @NonNull final String toolCode) {
        final Intent intent = new Intent(context, ToolDetailsActivity.class);
        intent.putExtra(EXTRA_TOOL, toolCode);
        context.startActivity(intent);
    }

    /* BEGIN lifecycle */

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Intent intent = getIntent();
        final Bundle extras = intent != null ? intent.getExtras() : null;
        if (extras != null) {
            mTool = extras.getString(EXTRA_TOOL, mTool);
        }

        // finish now if this activity is in an invalid state
        if (!validStartState()) {
            finish();
            return;
        }

        setContentView(R.layout.activity_generic_fragment_with_nav_drawer);
    }

    @Override
    protected void onSetupActionBar() {
        super.onSetupActionBar();
        setTitle("");
    }

    @Override
    protected void onStart() {
        super.onStart();
        loadInitialFragmentIfNeeded();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mEventBus.post(new AnalyticsScreenEvent(SCREEN_TOOL_DETAILS));
    }

    @Override
    public void onToolAdded() {
        finish();
    }

    @Override
    public void onToolRemoved() {
        finish();
    }

    /* END lifecycle */

    private boolean validStartState() {
        return mTool != null;
    }

    @MainThread
    private void loadInitialFragmentIfNeeded() {
        final FragmentManager fm = getSupportFragmentManager();

        // short-circuit if there is a currently attached fragment
        Fragment fragment = fm.findFragmentByTag(TAG_MAIN_FRAGMENT);
        if (fragment != null) {
            return;
        }

        // update the displayed fragment
        fm.beginTransaction()
                .replace(R.id.frame, ToolDetailsFragment.newInstance(mTool), TAG_MAIN_FRAGMENT)
                .commit();
    }

    @Override
    public void supportNavigateUpTo(@NonNull final Intent upIntent) {
        finish();
    }
}

File: app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsActivity.kt
package org.cru.godtools.ui.tooldetails

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.annotation.MainThread
import androidx.fragment.app.commit
import dagger.hilt.android.AndroidEntryPoint
import org.cru.godtools.R
import org.cru.godtools.activity.BasePlatformActivity
import org.cru.godtools.base.EXTRA_TOOL
import org.cru.godtools.base.ui.activity.BaseActivity
import org.cru.godtools.base.ui.util.getName
import org.cru.godtools.databinding.ActivityGenericFragmentWithNavDrawerBinding
import org.cru.godtools.model.Tool
import org.cru.godtools.model.Translation
import org.cru.godtools.tutorial.PageSet
import org.cru.godtools.tutorial.activity.TutorialActivity
import org.cru.godtools.tutorial.activity.startTutorialActivity
import org.cru.godtools.ui.tooldetails.analytics.model.ToolDetailsScreenEvent
import org.cru.godtools.util.openToolActivity

fun Activity.startToolDetailsActivity(toolCode: String) {
    startActivity(
        Intent(this, ToolDetailsActivity::class.java)
            .putExtras(BaseActivity.buildExtras(this))
            .putExtra(EXTRA_TOOL, toolCode)
    )
}

fun Activity.launchTrainingTips(
    tool: Tool?,
    type: Tool.Type,
    translation: Translation,
    isDiscovered: Boolean = false,
    force: Boolean
) {
    val code = tool?.code ?: return
    val locale = translation?.languageCode

    if (force || isDiscovered) openToolActivity(code, type, locale, showTips = true)
    else startTutorialActivity(PageSet.TIPS, Bundle().apply {
        putString(TutorialActivity.FORMAT_ARG_TOOL_NAME, translation.getName(tool).toString())
    }
    )
}

@AndroidEntryPoint
class ToolDetailsActivity : BasePlatformActivity<ActivityGenericFragmentWithNavDrawerBinding>() {
    // these properties should be treated as final and only set/modified in onCreate()
    private lateinit var tool: String

    // region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // finish now if we couldn't process the intent
        if (!processIntent()) finish()
    }

    override fun onSetupActionBar() {
        super.onSetupActionBar()
        title = ""
    }

    override fun onStart() {
        super.onStart()
        loadInitialFragmentIfNeeded()
    }

    override fun onResume() {
        super.onResume()
        eventBus.post(ToolDetailsScreenEvent(tool))
    }
    // endregion Lifecycle

    override fun inflateBinding() = ActivityGenericFragmentWithNavDrawerBinding.inflate(layoutInflater)

    /**
     * @return true if the intent was successfully processed, otherwise return false
     */
    private fun processIntent(): Boolean {
        tool = intent?.extras?.getString(EXTRA_TOOL) ?: return false
        return true
    }

    @MainThread
    private fun loadInitialFragmentIfNeeded() {
        with(supportFragmentManager) {
            if (primaryNavigationFragment != null) return

            commit {
                val fragment = ToolDetailsFragment(tool)
                replace(R.id.frame, fragment)
                setPrimaryNavigationFragment(fragment)
            }
        }
    }
}
